### DISCUSSION TITLE
Codegen to Generate schema from Remote GraphQL URL

### DISCUSSION THREAD

**Discussion by raghuraman-varadharajan** [Netflix/dgs-framework#1148] [Q&A] [answered]
Hello,

I am using the Hasura GraphQL engine that exposes the data from a Postgre SQL Database. I know we can create the schema and let the codegen generate the code for us which is cool.  

I am looking for a way given the  GraphQL endpoint could DGS automatically create both the schema and its corresponding codegen (Java Objects for respective schemas)?

**Comment by manymotes** [comment]
With the following build.gradle in a different module, I was able to point to any graphql schema and make the needed java classes:

```
buildscript {
    repositories {
        maven {
            url "https://plugins.gradle.org/m2/"
        }
    }
    dependencies {
        classpath "com.expediagroup:graphql-kotlin-gradle-plugin:5.3.2"
    }
}

plugins {
    id 'java'
    id 'com.expediagroup.graphql' version '5.3.2'
    id 'com.netflix.dgs.codegen' version '5.1.17'
}

generateJava {
    dependsOn graphqlIntrospectSchema
    typeMapping = [
            "DateTime"    : "java.time.OffsetDateTime",
            "Decimal"     : "java.math.BigDecimal"
    ]
    schemaPaths = ["somePath"] // List of directories containing schema files
    packageName = 'com.package.yourDir' // The package name to use to generate sources
    generateClient = true // Enable generating the type safe query API
}

group = 'com.pakcage'
version = '0.0.9'
sourceCompatibility = '17'

Properties properties = new Properties()
def credentialToken
def federatedUrl
def propertiesFile = project.rootProject.file('local.env')
if (propertiesFile.exists()) {
    properties.load(propertiesFile.newDataInputStream())
    credentialToken = properties.getProperty('someCred')
    federatedUrl = properties.getProperty('targetUrl')
} else {
    credentialToken = System.getenv('someCred')
    federatedUrl = System.getenv('targetUrl')
}


repositories {
    mavenCentral()
}

graphqlIntrospectSchema {
    endpoint = federatedUrl
    headers = ["securityHeader": someCred]
}

task cleanSchemaFile() {
    doFirst {
        graphqlIntrospectSchema
    }
}

test {
    useJUnitPlatform()
}

task compileKotlin {
    // noop
}
}```

**Reply by manymotes** [reply]
Maybe it would be smart to make a pr for a boiler plate module with this build.gradle

**Comment by manymotes** [comment, marked as answer]
Expedia docs:
https://opensource.expediagroup.com/graphql-kotlin/docs/plugins/gradle-plugin-tasks/
how to copy an external schema^^
Netflix docs:
https://netflix.github.io/dgs/generating-code-from-schema/#generating-query-apis-for-external-services
Netflix suggestion to create external graph code in another module^^

This solution was bought to you Expedia and Netflix, I would like to thank them for their great travel options and for making Stranger things respectively. lol, hope this helps!
